src/librustc_data_structures/profiling.rs

```diff
@@ -97,14 +97,27 @@ use std::time::{Duration, Instant};
 use measureme::{EventId, EventIdBuilder, SerializableString, StringId};
 use parking_lot::RwLock;
 
-/// MmapSerializatioSink is faster on macOS and Linux
-/// but FileSerializationSink is faster on Windows
-#[cfg(all(not(windows), not(target_arch = "wasm32")))]
-type SerializationSink = measureme::MmapSerializationSink;
-#[cfg(all(windows, not(target_arch = "wasm32")))]
-type SerializationSink = measureme::FileSerializationSink;
-#[cfg(target_arch = "wasm32")]
-type SerializationSink = measureme::ByteVecSink;
+cfg_if! {
+    if #[cfg(target_arch = "wasm32")] {
+        cfg_if! {
+            if #[cfg(target_os = "wasi")] {
+                type SerializationSink = measureme::FileSerializationSink;
+            } else {
+                type SerializationSink = measureme::ByteVecSink;
+            }
+        }
+    } else {
+        cfg_if! {
+            if #[cfg(windows)] {
+                /// FileSerializationSink is faster on Windows
+                type SerializationSink = measureme::FileSerializationSink;
+            } else {
+                /// MmapSerializatioSink is faster on macOS and Linux
+                type SerializationSink = measureme::MmapSerializationSink;
+            }
+        }
+    }
+}
```

Comment:
```suggestion
cfg_if! {
    if #[cfg(any(windows, target_os = "wasi"))] {
        /// FileSerializationSink is faster on Windows
        type SerializationSink = measureme::FileSerializationSink;
    } else if #[cfg(target_arch = "wasm32")] {
        type SerializationSink = measureme::ByteVecSink;
    } else {
        /// MmapSerializatioSink is faster on macOS and Linux
        type SerializationSink = measureme::MmapSerializationSink;
    }
}
```

Reply:
Whoops. Race condition :laughing:. I'd make `ByteVecSink` the fallback no?

Reply:
Maybe `FileSerializationSink` is the best fallback, as most systems support files.